mautic/mautic
# Fixed broken email A/B tests, improved email weights for campaign sent email, improved serving landing page A/B tests based on hits

Please answer the following questions. 

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | n
| Related user documentation PR URL | na
| Related developer documentation PR URL | na
| Issues addressed (#s or URLs) | #1520
| BC breaks? | n
| Deprecations? | n

**Note that all new features should have a related user and/or developer documentation PR in their respective repositories.** 

### Required
#### Description:

This PR does a few things.

1. A bug was introduced in 2.1 that caused A/B tests to not reset the variant date correctly which effectually broke A/B tests when it comes to determining what email to send.
2. The A/B test was not reset when a published variant was deleted so this is now fixed.
3. The A/B test was reset if an unpublished variant was updated
3. Improves support for campaigns. Since campaigns are currently processed one contact at a time, the same email was sent to all contacts processed during the same campaign trigger process. This PR changes it so that on each send through sendEmail, the weights are recalculated so that emails are sent in proportion to the A/B test's weight. 
4. Improves which page variant is displayed for landing page A/B tests based on variant weight

#### Steps to test this PR:
1. Run the unit tests
2. Create a new email with subject `email A`
3. Create an A/B test of that email and give 25% weight (choose any winning criteria) - change the subject to indicate `email B`
4. Create a second A/B test and give that email 25% and give a subject to indicate `email C`
5. Add 12 contacts to a segment
6. Create a campaign that sends the email to that segment and trigger
7. 6 of email A, and 3 each of email B and C should have been sent.
8. Create a new email A, B, and C as a segment email
9. Send the email to the segment with the 12 contacts
10. 6 of email A, and 3 each of email B and C should have been sent.

For landing pages
1. Run the unit tests
2. Create a landing page A
3. Create A/B tests B and C and add 25% weight each
4. In an anonymous browser, hit up the public URL - should show page A
5. Clear cookies and hit it again - should show page B
6. Again - page C
7. Again - page A

### As applicable
#### Steps to reproduce the bug:
1. Repeat the process above - the campaign will result in a single email going to all 12 contacts; if the campaign events are wiped and the campaign triggered again, all 12 would likely receive one of the variants
